### ISSUE TITLE
A11y 1.1 - Permalinks missing accessible name

### ISSUE THREAD

**Issue by anevins12** [mattgarrish/epub-revision#1827]
The permalinks used in list items have no accessible name and won't be identified to assistive technologies (AT). I don't think the visual indication of the link being a greater than symbol conveys that this is a permalink to non-AT users either. This is the markup and CSS adds the greater than symbol, which is announced by some AT:
```
<a href="#confreq-schema-accessMode" class="hidden-reveal"><span>&nbsp;</span></a>
```

In other cases the section character is used, but its accessible name is not descriptive (it is the § character):
```
<a class="self-link" aria-label="§" href="#sec-wcag-eval-page-pub"></a>
```

Whereas in comparison to the Recommendation on WCAG 2.1, permalinks use `aria-label` to provide a descriptive name and the `title` attribute for that same information to non AT users.
```
<span class="permalink">
  <a href="#background-on-wcag-2" aria-label="Permalink for 0.1 Background on WCAG 2" title="Permalink for 0.1 Background on WCAG 2"><span>§</span></a>
</span>
```

**Comment by anevins12** [comment]
Would you consider leaving the bullet style of the list so that it retains its semantics in VoiceOver?

**Comment by mattgarrish** [comment]
> I don't think the visual indication of the link being a greater than symbol conveys that this is a permalink to non-AT users either.

This is a relic of the IDPF specifications. I don't believe W3C adds permalinks to list items, or at least I don't see how it can be handled in respec (there's some custom javascript code that runs in a post-processing step to add these).

I wonder if we really need the permalinks on the conformance requirements, as I can't think of an easy way to add context to all the bullets, but this would potentially affect all our specifications. Any thoughts on this @iherman @dauwhe @wareid @shiestyle ?

> In other cases the section character is used, but its accessible name is not descriptive (it is the § character):

Yes, these are coming from the respec code. The issue affects all W3C specifications including the current draft of WCAG. They must have patched the source for the REC document if there are different attributes.

This case is probably less difficult to handle in a post-processing step, though, since it shouldn't be too hard to grab the text content of the heading and repurpose it into the attributes WCAG uses. I'll see if I can wire up something similar.

A better solution, though, would be to get this integrated into respec, or whoever is responsible for the "addPermalinks" option, so it's not an issue everywhere. (I can't find that function documented in the respec docs, so maybe this is specific to W3C?)

**Comment by iherman** [comment]
> > I don't think the visual indication of the link being a greater than symbol conveys that this is a permalink to non-AT users either.
> 
> This is a relic of the IDPF specifications. I don't believe W3C adds permalinks to list items, or at least I don't see how it can be handled in respec (there's some custom javascript code that runs in a post-processing step to add these).

I do not think respec has any tool for this

> 
> I wonder if we really need the permalinks on the conformance requirements, as I can't think of an easy way to add context to all the bullets, but this would potentially affect all our specifications. Any thoughts on this @iherman @dauwhe @wareid @shiestyle ?

Having a permalink on conf. requirements may be very handy for test writing. There is a metadata item we use in each test that is supposed to link back to the 'place' in the spec that this test applies to. I wonder whether using a permalink to a conformance statement would make things easier. 

That being said, @mattgarrish, if you think it would make thing more complicated editorially then I do not mind dropping this.

(I did not see these conformance clause permalink in the /TR version, actually...)

**Comment by iherman** [comment]
> A better solution, though, would be to get this integrated into respec, or whoever is responsible for the "addPermalinks" option, so it's not an issue everywhere. (I can't find that function documented in the respec docs, so maybe this is specific to W3C?)

Absolutely. If adding the aria attribute and the title for the permalink is something that WCAG would require, it should be handled by respec. The proper way is, probably, to raise a feature request in https://github.com/w3c/respec/issues  (but the backlog is probably large, so we may have to keep to the script that you have in #1831 for the time being...)

Cc @marcoscaceres @sidvishnoi

**Comment by mattgarrish** [comment]
> There is a metadata item we use in each test that is supposed to link back to the 'place' in the spec that this test applies to.

I wouldn't remove the IDs from the list items, so you can still do this although it makes getting the ID a little harder if you're going by the rendered document and not the source.

Beyond how to get meaningful names, what bothers me about the lists is that the ones we've kept are detached from their original purpose. Prior to this revision, we had distinct sections called "Content Conformance" and "Reading System Conformance" where these lists occurred. After the content reshuffling, the remaining lists have lost their context, so whether a list has permalinks for each bullet seems a bit random now.

The only thought I've had to keep the visible permalinks would be to use an incrementing counter to number them, so the value might be "Permalink for conformance statement 1". It's not terribly enlightening, though. We could maybe generate numbers at the start of the bullets for better context (e.g., "[1] blah blah blah"), but then the randomness of these lists probably becomes even more apparent.

**Comment by iherman** [comment]
> > There is a metadata item we use in each test that is supposed to link back to the 'place' in the spec that this test applies to.
> 
> I wouldn't remove the IDs from the list items, so you can still do this although it makes getting the ID a little harder if you're going by the rendered document and not the source.

Yep.

(B.t.w. there is handy extension for mozilla and chrome called 'fragment' that may be of a great help.)

> 
> Beyond how to get meaningful names, what bothers me about the lists is that the ones we've kept are detached from their original purpose. Prior to this revision, we had distinct sections called "Content Conformance" and "Reading System Conformance" where these lists occurred. After the content reshuffling, the remaining lists have lost their context, so whether a list has permalinks for each bullet seems a bit random now.
> 
> The only thought I've had to keep the visible permalinks would be to use an incrementing counter to number them, so the value might be "Permalink for conformance statement 1". It's not terribly enlightening, though. We could maybe generate numbers at the start of the bullets for better context (e.g., "[1] blah blah blah"), but then the randomness of these lists probably becomes even more apparent.

I would say: nuke them! :-)

**Comment by marcoscaceres** [comment]
The code that adds the paragraph markers is here:

https://github.com/w3c/respec/blob/ba47f7fc51197f2665c07fb824b8a7782c88d0ae/src/core/id-headers.js#L15-L25

It's super simple and would love any help to improve it with an `aria-label` or whatever y'all suggest.

**Comment by mattgarrish** [comment]
> It's super simple and would love any help to improve it with an `aria-label` or whatever y'all suggest.

The challenge I had was figuring out what gets permalinks. From the EPUB specs, I found sections, appendixes, examples, and webidl instances all get permalinked, but I don't know if there are other things that could get these. Also, disambiguating which you are in requires inspecting what kind of parent the link has and whether there is a secno element you can grab.

The WebIDL instances are potentially the most problematic, as they don't have unique titles. I kind of lucked out with EPUB as there is only one instance, so calling it "Permalink for WebIDL" ended up unique enough, but we'd need something better for a general processing step.

The code I ended up with worked for what I was finding, but I'm not sure how robust it is. The code is here: https://github.com/w3c/epub-specs/blob/main/epub33/common/js/permalink-a11y.js which maybe could be used as a basis for updating the referenced function (except it's post-processing the markup, so would have to be modified).

I wonder how best we should do this, though, since the accessibility implications suggest this should probably be looked at by the APA WG. I didn't follow the same pattern as WCAG of using the whole heading as the accessible name, but shortened the attributes down so they followed the pattern "Permalink for Section/Appendix/Example #".
